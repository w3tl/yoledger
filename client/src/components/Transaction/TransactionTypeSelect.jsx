import React from 'react';
import PropTypes from 'prop-types';
import { withLabel } from '../Elements/utils';

class TransactionTypeSelect extends React.Component {
  static propTypes = {
    name: PropTypes.string,
    value: PropTypes.oneOf(['income', 'expense']),
    onChange: PropTypes.func,
  }

  constructor(props) {
    super(props);
    this.state = {
      value: props.value,
    };
  }

  handleChange = value => () => {
    const { onChange } = this.props;
    this.setState({
      value,
    });
    onChange({ target: { value } });
  }

  render() {
    const { name } = this.props;
    const { value } = this.state;

    return (
      <div>
        <input type="radio" name={name} value="income" id="income" checked={value === 'income'} onChange={this.handleChange('income')} />
        <label htmlFor="income">Income</label>
        <input type="radio" name={name} value="expense" id="expense" checked={value === 'expense'} onChange={this.handleChange('expense')} />
        <label htmlFor="expense">Expense</label>
      </div>);
  }
}

TransactionTypeSelect.defaultProps = {
  name: 'switch',
  value: 'expense',
  onChange: () => {},
};

export default withLabel(TransactionTypeSelect);
