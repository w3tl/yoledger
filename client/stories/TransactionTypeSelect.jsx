import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { TransactionTypeSelect } from '../src/components';

storiesOf('TransactionTypeSelect', module)
  .add('default', () => <TransactionTypeSelect />)
  .add('with label', () => <TransactionTypeSelect label="Select" />)
  .add('with value', () => <TransactionTypeSelect value="income" onChange={action('Value changed')} />);
