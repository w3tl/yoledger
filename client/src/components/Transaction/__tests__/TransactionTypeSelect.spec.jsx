import React from 'react';
import { configure, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import TransactionTypeSelect from '../TransactionTypeSelect';

configure({ adapter: new Adapter() });

describe('TransactionTypeSelect', () => {
  it('should render correctly', () => {
    const output = mount(
      <TransactionTypeSelect value="income" />,
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('should handle the input event', () => {
    const onChange = jest.fn();
    const output = mount(
      <TransactionTypeSelect onChange={onChange} value="income" />,
    );
    output.find('#expense').simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledWith({ target: { value: 'expense' } });
    expect(output.find('#expense').prop('checked')).toBeTruthy();
    expect(output.find('#income').prop('checked')).toBeFalsy();
  });
});
