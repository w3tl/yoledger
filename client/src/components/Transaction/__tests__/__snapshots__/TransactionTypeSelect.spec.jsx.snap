// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionTypeSelect should render correctly 1`] = `
"<Component value=\\"income\\">
  <TransactionTypeSelect id=\\"\\" value=\\"income\\" name=\\"switch\\" onChange={[Function: onChange]}>
    <div>
      <input type=\\"radio\\" name=\\"switch\\" value=\\"income\\" id=\\"income\\" checked={true} onChange={[Function]} />
      <label htmlFor=\\"income\\">
        Income
      </label>
      <input type=\\"radio\\" name=\\"switch\\" value=\\"expense\\" id=\\"expense\\" checked={false} onChange={[Function]} />
      <label htmlFor=\\"expense\\">
        Expense
      </label>
    </div>
  </TransactionTypeSelect>
</Component>"
`;
